packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts

```diff
@@ -0,0 +1,186 @@
+*--------------------------------------------------------------------------------------------*/
+import { ExtractionClient, ExtractorState, MappingsClient, REPORTING_BASE_PATH, ReportsClient } from "@itwin/insights-client";
+import type { ReportMapping } from "@itwin/insights-client";
+import { generateUrl, handleError } from "./utils";
+import type { CreateTypeFromInterface } from "./utils";
+import type {
+  GetSingleIModelParams,
+} from "@itwin/imodels-client-management";
+import { Constants, IModelsClient } from "@itwin/imodels-client-management";
+import { AccessTokenAdapter } from "@itwin/imodels-access-frontend";
+import type { ReportsApiConfig } from "../context/ReportsApiConfigContext";
+import { ExtractionStates } from "./ExtractionStatus";
+
+export type ReportMappingType = CreateTypeFromInterface<ReportMapping>;
+
+export type ReportMappingAndMapping = ReportMappingType & {
+  mappingName: string;
+  mappingDescription: string;
+  iModelName: string;
+};
+
+export default class BulkExtractor {
+  private _reportRunIds = new Map<string, string[]>();
+  private _reportsClientApi: ReportsClient;
+  private _mappingsClientApi: MappingsClient;
+  private _iModelsClient: IModelsClient;
+  private _extractionClientApi: ExtractionClient;
+  private _accessToken: Promise<string>;
+  private _iModelNames = new Map<string, string>();
+
+  constructor(apiConfig: ReportsApiConfig) {
+    const url = generateUrl(REPORTING_BASE_PATH, apiConfig.baseUrl);
+    this._reportsClientApi = new ReportsClient(url);
+    this._mappingsClientApi = new MappingsClient(url);
+    this._extractionClientApi = new ExtractionClient(url);
+    this._iModelsClient = new IModelsClient({
+      api: { baseUrl: generateUrl(Constants.api.baseUrl, apiConfig.baseUrl) },
+    });
+    this._accessToken = apiConfig.getAccessToken();
+  }
+
+  public async getStates(reportIds: string[]): Promise<Map<string, ExtractionStates>> {
+    const stateByReportId = new Map<string, ExtractionStates>();
+    const stateByRunId = new Map<string, ExtractorState>();
+
+    for (const reportId of reportIds) {
+      const runs = this._reportRunIds.get(reportId);
+      if (!runs) {
+        stateByReportId.set(reportId, ExtractionStates.None);
+        continue;
+      }
+
+      const states: ExtractorState[] = [];
+      for (const runId of runs) {
+        const state = stateByRunId.get(runId);
+        if (state) {
+          states.push(state);
+        } else {
+          const runState = await this.getSingleState(runId, await this._accessToken);
+          states.push(runState);
+          stateByRunId.set(runId, runState);
+        }
+      }
+      const finalState = this.getFinalState(states);
+      stateByReportId.set(reportId, finalState);
+    }
+    return stateByReportId;
+  }
+
+  public clearJob(reportId: string) {
+    this._reportRunIds.delete(reportId);
+  }
+
+  private getFinalState(states: ExtractorState[]) {
```

Comment:
Make `getFinalState` static since it does not use any of the fields of the object.